keras_list_mapper/keras_list_mapper.py

```diff
@@ -111,73 +114,75 @@ def __call__(self, inputs, *args, **kwargs):
-                spec = getattr(inp, '_type_spec', None)
-            if spec is None or not isinstance(spec, tf.RaggedTensorSpec) \
-                    or ix in self.batch_inputs:
+        has_ragged = False
+        input_shape = nest.map_structure(
+            lambda x: tf.TensorShape(K.int_shape(x)), inputs)
+        batch_inputs = nest.flatten(self._batch_inputs(input_shape))
+        for batch_input, inp in zip(batch_inputs, inputs_list):
+            if batch_input:
                 inner_inputs.append(inp)
                 continue
-            count += 1
-            # TODO(roque): remove _shape after v2.3
-            if hasattr(spec, 'shape'):
-                nshape = tf.TensorShape(spec.shape[:1] + spec.shape[2:])
-            else:
-                nshape = tf.TensorShape(spec._shape[:1] + spec._shape[2:])
-            assert isinstance(spec, tf.RaggedTensorSpec)
+            has_ragged = True
+            nshape = tf.TensorShape(inp.shape[:1] + inp.shape[2:])
+            assert isinstance(inp.type_spec, tf.RaggedTensorSpec)
             is_ragged = self.mapper_supports_ragged_inputs
             pl = K.placeholder(shape=nshape, dtype=inp.dtype, ragged=is_ragged)
             inner_inputs.append(pl)
 
-        if not count:
-            raise ValueError(
-                'ListMapper inputs has no RaggedTensor')
+        if not has_ragged:
+            raise ValueError('ListMapper inputs has no RaggedTensor')
 
         if len(inner_inputs) == 1:
             inner_inputs = inner_inputs[0]
         outputs = self.layer(inner_inputs, *args, **kwargs)
 
-        # tracing the function while saving the graph
-        if call_context().saving:
-            if self.state_shape is not None:
-                outputs = outputs[1:]
-            elif not isinstance(outputs, (list, tuple)):
-                outputs = [outputs]
-            outputs = [x._to_placeholder() if K.is_keras_tensor(x) else x
-                       for x in outputs]
-            return outputs
+        if self.state_shape is not None:
+            outputs = outputs[1:]
+            if len(outputs) == 1:
+                outputs = outputs[0]
 
         if not self._built_from_signature:
-            if self.state_shape is not None:
-                outputs = outputs[1]
-            # TODO(roque): remove _type_spec
-            self._output_signature = getattr(
-                outputs, 'type_spec', tf.TensorSpec(outputs.shape))
+            self._output_signatures = nest.map_structure(
+                lambda x: x.type_spec, outputs
+            )
             self._built_from_signature = True
 
-        super().build(inputs)
+        # Build layer and inner layer
+        self.build(input_shape)
+        output_shapes = self.compute_output_shape(input_shape)
+
+        def replace_shape(tensor, shape: tf.TensorShape):
+            return K.placeholder(shape, dtype=tensor.dtype, ragged=True)
+        outputs = nest.map_structure(replace_shape, outputs, output_shapes)
```

Comment:
Is this required ? My concern is that the tensor outputs from calling the inner layer are discarded.
While working on this previously keras was rather keen on having the outputs of the layer connected to the next layer. That is the reason _set_connectivity_metadata is called... without it the graph is considered to be disconnected. I don't think the outputs should be overriden.

I understand that we want to make the outputs a ragged tensor but it would be nice to find a way to do that without using a placeholder.

Reply:
My comment on `_set_connectivity_metadata` is wrong, it is indeed responsible for connecting the graph.

Regarding the connection with the inner layer: it seems that wrappers are not connected to their inner layer. In order for a graph to be properly connected, inputs and outputs of each layer must be connected. Since the inner layer is not part of the graph this doesn't seem to be a problem.

Example:
```python
inputs = layers.Input((None, 1), ragged=True)
td = layers.TimeDistributed(layers.Dense(1))(inputs)
lstm = layers.LSTM(1)(td)
model = models.Model(inputs, lstm)
assert len(model.layers[-2]._outbound_nodes) == 1
assert len(model.layers[-2].layer._outbound_nodes) == 0
```
```python
inputs = layers.Input((None, 1), ragged=True)
lm = ListMapper(layers.Dense(1))(inputs)
lstm = layers.LSTM(1)(lm)
model = models.Model(inputs, lstm)
assert len(model.layers[-2]._outbound_nodes) == 1
assert len(model.layers[-2].layer._outbound_nodes) == 0
```

The best solution of course would be not to override `__call__` and use `call` also when building the graph. But with the current implementation this is still not possible. [Dynamic layers](https://github.com/tensorflow/tensorflow/blob/ec5ff5df834bcd32941ffac649950558f2adea1f/tensorflow/python/keras/engine/base_layer.py#L864) are built in a similar fashion.